File: apps/android/core/src/main/kotlin/dev/pigeonpost/core/MailboxOrder.kt
package dev.pigeonpost.core

import java.util.Locale

/** A stable picker order, independent of the inbox currently being read. */
fun orderedMailboxes(mailboxes: List<Mailbox>, username: String?): List<Mailbox> {
    data class Row(val index: Int, val mailbox: Mailbox, val path: String, val namespace: String, val named: Boolean, val root: Boolean)
    val rows = mailboxes.mapIndexed { index, mailbox ->
        val path = mailbox.handle.orEmpty().trim().lowercase(Locale.ROOT)
        val parts = path.trim('/').split('/')
        val named = path.startsWith('/') && parts.first().isNotEmpty() && parts.first() != "k"
        val root = named && (parts.size == 1 || (parts.size == 2 && parts.last() == "main"))
        Row(index, mailbox, path.ifEmpty { mailbox.address }, if (named) parts.first() else "", named, root)
    }
    val user = username.orEmpty().trim().trim('/').lowercase(Locale.ROOT)
    val roots = rows.filter { it.root }
    val primary = roots.firstOrNull { user.isNotEmpty() && it.namespace == user } ?: roots.firstOrNull()
    return rows.sortedWith(compareBy<Row> {
        when { it.index == primary?.index -> 0; it.root -> 1; it.named -> 2; else -> 3 }
    }.thenBy { it.namespace }.thenBy { it.path }.thenBy { it.index }).map { it.mailbox }
}

File: apps/ios/Pigeonpost/Views/MailboxOrder.swift
import Foundation

/// Picker order is independent of the mailbox currently being read.
enum MailboxOrder {
    static func sorted(_ mailboxes: [Mailbox], primaryNamespace: String, username: String?) -> [Mailbox] {
        let rows = mailboxes.enumerated().map { index, mailbox in
            let path = (mailbox.handle ?? "").trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
            let parts = path.split(separator: "/")
            let named = path.hasPrefix("/") && !parts.isEmpty && parts.first != "k"
            let root = named && (parts.count == 1 || (parts.count == 2 && parts.last == "main"))
            return (index: index, mailbox: mailbox, path: path,
                    namespace: named ? String(parts[0]) : "", named: named, root: root)
        }
        func namespace(_ value: String) -> String {
            value.trimmingCharacters(in: CharacterSet(charactersIn: "/").union(.whitespacesAndNewlines)).lowercased()
        }
        let roots = rows.filter(\.root)
        let username = namespace(username ?? "")
        let primary = roots.first { !username.isEmpty && $0.namespace == username }
            ?? roots.first { $0.namespace == namespace(primaryNamespace) }
            ?? roots.first
        func rank(_ index: Int, _ named: Bool, _ root: Bool) -> Int {
            if index == primary?.index { return 0 }
            return root ? 1 : named ? 2 : 3
        }
        return rows.sorted { a, b in
            let ar = rank(a.index, a.named, a.root), br = rank(b.index, b.named, b.root)
            if ar != br { return ar < br }
            if a.namespace != b.namespace { return a.namespace < b.namespace }
            if a.path != b.path { return a.path < b.path }
            return a.index < b.index
        }.map(\.mailbox)
    }
}
